src/librustc_mir/dataflow/generic/tests.rs

```diff
@@ -0,0 +1,297 @@
+    /// at a `Call` terminator that can return. Returns `None` for `BlockStart`.
+    fn effect(&self, body: &mir::Body<'_>) -> Option<usize> {
+        use SeekTarget::*;
+
+        let idx = match *self {
+            BlockStart(_) => return None,
+
+            AfterAssumeCallReturns(loc) if is_call_terminator_non_diverging(body, loc)
+                => loc.statement_index * 2 + 2,
+
+            Before(loc) => loc.statement_index * 2,
+            After(loc) | AfterAssumeCallReturns(loc) => loc.statement_index * 2 + 1,
+        };
+
+        assert!(idx < BASIC_BLOCK_OFFSET, "Too many statements in basic block");
+        Some(idx)
+    }
+
+    /// An iterator over all possible `SeekTarget`s in a given block in order, starting with
+    /// `BlockStart`.
+    ///
+    /// This includes both `After` and `AfterAssumeCallReturns` for every `Location`.
+    fn iter_in_block(body: &mir::Body<'_>, block: BasicBlock) -> impl Iterator<Item = Self> {
+        let statements_and_terminator = (0..=body[block].statements.len())
+            .flat_map(|i| (0..3).map(move |j| (i, j)))
+            .map(move |(i, kind)| {
+                let loc = Location { block, statement_index: i };
+                match kind {
+                    0 => SeekTarget::Before(loc),
+                    1 => SeekTarget::After(loc),
+                    2 => SeekTarget::AfterAssumeCallReturns(loc),
+                    _ => unreachable!(),
+                }
+            });
+
+        std::iter::once(SeekTarget::BlockStart(block))
+            .chain(statements_and_terminator)
+    }
+}
+
+/// A mock dataflow analysis for testing `ResultsCursor`.
+///
+/// `Dummy` assigns a unique, monotonically increasing index to each possible cursor position as
+/// well as one to each basic block. Instead of being iterated to fixpoint, `Dummy::entry_sets` is
+/// used to construct the entry set to each block such that the dataflow state that should be
+/// observed by `ResultsCursor` is unique for every location (see `expected_state_at`).
+struct Dummy<'tcx> {
+    body: &'tcx mir::Body<'tcx>,
+}
+
+impl Dummy<'tcx> {
+    fn entry_sets(&self) -> IndexVec<BasicBlock, BitSet<usize>> {
+        let empty = BitSet::new_empty(self.bits_per_block(self.body));
+        let mut ret = IndexVec::from_elem(empty, &self.body.basic_blocks());
+
+        for (bb, _) in self.body.basic_blocks().iter_enumerated() {
+            ret[bb].insert(BASIC_BLOCK_OFFSET + bb.index());
+        }
+
+        ret
+    }
+
+    /// Returns the expected state at the given `SeekTarget`.
+    ///
+    /// This is the union of index of the target basic block, the index assigned to the of the
+    /// target statement or terminator, and the indices of all preceding statements in the MIR.
+    ///
+    /// For example, the expected state when calling
+    /// `seek_before(Location { block: 2, statement_index: 2 })` would be `[102, 0, 1, 2, 3, 4]`.
```

Comment:
I spent a little while reading over the dummy_body()  cfg construction code and tried to understand why that control-flow graph would yield the results `[102, 0, 1, 2, 3, 4]`, but at some point I decided that if I have to work this hard to figure it out, then it needs to be explained more explicitly in the documentation here. In short, I don't know why I should expect the aforementioned output for that case. (e.g. Where does the 102 come from?)